emcc.py

```diff
@@ -3279,6 +3279,20 @@ def generate_minimal_runtime_html(target, options, js_target, target_basename,
   if re.search(r'{{{\s*SCRIPT\s*}}}', shell):
     exit_with_error('--shell-file "' + options.shell_path + '": MINIMAL_RUNTIME uses a different kind of HTML page shell file than the traditional runtime! Please see $EMSCRIPTEN/src/shell_minimal_runtime.html for a template to use as a basis.')
 
+  # Depending on whether streaming Wasm compilation is enabled or not, the minimal sized code to download Wasm looks a bit different.
+  # Expand {{{ DOWNLOAD_WASM }}} block from here (if we added #define support, this could be done in the template directly)
+  if shared.Settings.MINIMAL_RUNTIME_STREAMING_WASM_COMPILATION:
+    if shared.Settings.MIN_SAFARI_VERSION != shared.Settings.TARGET_NOT_SUPPORTED or shared.Settings.ENVIRONMENT_MAY_BE_NODE:
+      # https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/compileStreaming
+      # In Safari and Node.js, WebAssembly.compileStreaming() is not supported, in which case fall back to regular download.
+      download_wasm = "WebAssembly.compileStreaming ? WebAssembly.compileStreaming(fetch('{{{ TARGET_BASENAME }}}.wasm')) : b('{{{ TARGET_BASENAME }}}.wasm')"
```

Comment:
`instantiateStreaming` is even more efficient - maybe worth going straight to that?

Reply:
Actually, I strongly disagree here. (unless my understanding of the cost model is very off here) `instantiateStreaming` is worse because it requires one to download, parse and run `a.js` first before being able to begin fetch+instantiateStreaming of `a.wasm` at all.

That is, my understanding is that Wasm compile is the majority of the work compared to Wasm instantiation. (compile is the actual module parsing, validation and compilation to native bytecode - instantiation is just more or less constant time execution env setup?)

If so, being able to kick off fetch+compile of wasm as early on on the page as possible is most important.

Another minor advantage of this is that this makes the main thread startup analogous to how pthreads instantiate as well, i.e. the ready made module flows in to both main thread code and pthreads code to get instantiated.

Reply:
Yeah, good point, it's possible that starting the wasm compilation before getting the .js file is worth it. But if the .js file is small, then instantiateStreaming should beat compileStreaming AFAIK, as there is some VM benefit to doing it all at once (in particular it can see the wasm imports before compiling). I haven't measured this though.

Reply:
Hmm, interesting.. this may warrant comparative profiling at some point. I'll put this on the list to check.